File: packages/sdk-python/src/driftstack/resources/recipes.py
"""Recipes resource — /v1/recipes (AI-B4).

Mirrors the TypeScript RecipesResource. Server registers the routes as
503 ``FeatureUnavailable`` stubs until both ``recipesRepo`` and
``agentSessionsRepo`` are wired in AppDeps; SDK surface is stable so
consumers compile ahead of time.

Surface: ``create`` + ``list`` + ``get`` + ``delete`` (the
read/management path was pulled forward from the v1.1 D2/D3 defer —
V-530.I/.J). Recipe EXECUTION stays v1.1 (gated on the harness executor).
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Iterator
from typing import Any
from urllib.parse import quote, urlencode

from driftstack.http import AsyncHttpClient, HttpClient
from driftstack.pagination import aiterate_paginated, iterate_paginated
from driftstack.resources._common import coerce_body


def _encode_query(query: dict[str, Any]) -> str:
    items: list[tuple[str, str]] = []
    for key, value in query.items():
        if value is None:
            continue
        items.append((key, str(value)))
    return urlencode(items)


class RecipesResource:
    """Synchronous AI-B4 recipe library."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        *,
        agent_session_id: str,
        label: str,
        description: str | None = None,
    ) -> dict[str, Any]:
        """Snapshot a finished agent_session's intent_log + transcript.

        Returns the inserted ``Recipe`` payload (id + account_id +
        agent_session_id + label + description + intent_count +
        timestamps).

        Server-side: cross-account access on ``agent_session_id``
        returns 404 (not 403) by design — existence isn't leaked.
        """
        body: dict[str, Any] = {
            "agent_session_id": agent_session_id,
            "label": label,
        }
        if description is not None:
            body["description"] = description
        return self._http.request("POST", "/v1/recipes", json_body=coerce_body(body))

    def list(self, *, limit: int | None = None, cursor: str | None = None) -> dict[str, Any]:
        """List the account's recipes, newest first. Cursor-paginated."""
        qs = _encode_query({"limit": limit, "cursor": cursor})
        path = "/v1/recipes" + (f"?{qs}" if qs else "")
        return self._http.request("GET", path)

    def iterate(self, *, limit: int | None = None) -> Iterator[dict[str, Any]]:
        """Lazily walk every recipe, handling cursor handoff."""

        def fetch_page(cursor: str | None) -> dict[str, Any]:
            return self.list(limit=limit, cursor=cursor)

        return iterate_paginated(fetch_page)

    def get(self, recipe_id: str) -> dict[str, Any]:
        """Fetch one recipe with its public ``intent_log``.

        Sensitive type steps retain their selector and ``sensitive`` marker but
        omit the optional value. Exact replay values stay encrypted server-side.

        404 if missing or owned by another account (existence not leaked).
        """
        return self._http.request("GET", f"/v1/recipes/{quote(recipe_id, safe='')}")

    def delete(self, recipe_id: str) -> None:
        """Delete a recipe. 404 if missing or owned by another account."""
        self._http.request("DELETE", f"/v1/recipes/{quote(recipe_id, safe='')}")

    def suggest(self, agent_session_id: str) -> dict[str, Any]:
        """Doc-132 §5.2 (recipe auto-generation) v1.0 slice.

        Fetch a deterministic label/description suggestion derived from
        the session's own intent_log (same assembly ``create`` uses).
        Read-only; safe to call speculatively before deciding to save.
        Returns ``suggested_label`` / ``suggested_description`` /
        ``intent_count``.
        """
        return self._http.request(
            "GET", f"/v1/agent-sessions/{quote(agent_session_id, safe='')}/recipe-suggestion"
        )


class AsyncRecipesResource:
    """Async AI-B4 recipe library."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def create(
        self,
        *,
        agent_session_id: str,
        label: str,
        description: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "agent_session_id": agent_session_id,
            "label": label,
        }
        if description is not None:
            body["description"] = description
        return await self._http.request("POST", "/v1/recipes", json_body=coerce_body(body))

    async def list(self, *, limit: int | None = None, cursor: str | None = None) -> dict[str, Any]:
        qs = _encode_query({"limit": limit, "cursor": cursor})
        path = "/v1/recipes" + (f"?{qs}" if qs else "")
        return await self._http.request("GET", path)

    def iterate(self, *, limit: int | None = None) -> AsyncIterator[dict[str, Any]]:
        """Async variant of :meth:`RecipesResource.iterate`."""

        async def fetch_page(cursor: str | None) -> dict[str, Any]:
            return await self.list(limit=limit, cursor=cursor)

        return aiterate_paginated(fetch_page)

    async def get(self, recipe_id: str) -> dict[str, Any]:
        return await self._http.request("GET", f"/v1/recipes/{quote(recipe_id, safe='')}")

    async def delete(self, recipe_id: str) -> None:
        await self._http.request("DELETE", f"/v1/recipes/{quote(recipe_id, safe='')}")

    async def suggest(self, agent_session_id: str) -> dict[str, Any]:
        """Async variant of :meth:`RecipesResource.suggest`."""
        return await self._http.request(
            "GET", f"/v1/agent-sessions/{quote(agent_session_id, safe='')}/recipe-suggestion"
        )

File: apps/server/src/services/recipes.ts
// AI-B4 — recipes persistence. A recipe is a snapshot of a finished
// agent_session's intent_log + transcript so the customer can replay
// the same flow later via the SDK without re-paying the LLM
// decomposition cost.
//
// Surface: create + list + getById + deleteById (the read/management
// path was pulled forward from the v1.1 D2/D3 defer — V-530.I/.J).
// Recipe EXECUTION stays v1.1 (gated on the harness-wired executor).
//
// Migration: 0044_recipes.sql. Schema follows the same text-PK +
// jsonb-payload pattern as agent_sessions.

import { randomUUID } from 'node:crypto';
import type { AccountTier } from '@driftstack/api-types';
import type { AgentIntent, TranscriptEntry } from './agent-decomposer.js';

/**
 * Security sweep #7 — how many recipes an account keeps, per plan.
 *
 * Every recipe stores its own encrypted copy of the source session's transcript
 * (up to 1 MiB, and ciphertext does not compress), so the count bounds the bytes
 * one account can put into the database: 10 on free is at most ~10 MiB. Paid
 * plans keep more, in rough proportion to their profile allowance.
 */
export const RECIPES_PER_TIER: Readonly<Record<AccountTier, number>> = {
  free: 10,
  solo_manual: 50,
  team_manual: 200,
  agency_manual: 500,
  api_starter: 100,
  api_builder: 500,
  api_scale: 1000,
  enterprise: 2000,
};

/** The recipe cap for a plan. Enforced by `RecipesRepo.createIfUnderLimit`, under a lock. */
export function recipeLimitFor(tier: AccountTier): number {
  return RECIPES_PER_TIER[tier];
}

export interface RecipeRecord {
  /** `rec_<uuid>` id; minted by the repo on create. */
  id: string;
  accountId: string;
  /**
   * Source agent-session this recipe was snapshotted from. NULLABLE
   * because agent sessions may be deleted later but the recipe
   * row survives — ON DELETE SET NULL preserves the recipe while
   * dropping the dangling reference.
   */
  agentSessionId: string | null;
  label: string;
  description: string | null;
  /**
   * Captured plan — the ordered AgentIntent sequence the agent
   * executed. The replay path (v1.1) iterates this in order.
   */
  intentLog: ReadonlyArray<AgentIntent>;
  /**
   * Captured transcript at snapshot time — useful context for "what
   * did I ask?" when the customer revisits the recipe months later.
   */
  transcriptSnapshot: ReadonlyArray<TranscriptEntry>;
  createdAt: Date;
  updatedAt: Date;
}

export interface CreateRecipeArgs {
  accountId: string;
  /** Source agent session id; the service reads its transcript +
   *  intent log at snapshot time. Pass NULL when the customer
   *  composes a recipe out of band (not the v1.0 path; the route
   *  layer always passes a real id). */
  agentSessionId: string | null;
  label: string;
  description?: string;
  intentLog: ReadonlyArray<AgentIntent>;
  transcriptSnapshot: ReadonlyArray<TranscriptEntry>;
}

export interface ListRecipesArgs {
  accountId: string;
  /** Page size (default 50, max 100 — clamped by the repo). */
  limit?: number;
  /** Opaque cursor = the id of the last recipe on the prior page. */
  cursor?: string;
}

export interface ListRecipesPage {
  data: RecipeRecord[];
  hasMore: boolean;
  /** The id to pass as the next `cursor`, or null when the page is the last. */
  nextCursor: string | null;
}

/** What {@link RecipesRepo.createIfUnderLimit} decided. */
export type CreateRecipeOutcome =
  /** A new recipe was stored. */
  | { kind: 'created'; record: RecipeRecord }
  /** The same session, label and description were already saved: that recipe, unchanged. */
  | { kind: 'existing'; record: RecipeRecord }
  /** The session is already saved as `recipeId` under another label or description. */
  | { kind: 'session_already_saved'; recipeId: string }
  /** The account already holds `current` recipes, at or over the limit. */
  | { kind: 'limit_reached'; current: number };

export interface RecipesRepo {
  /**
   * Snapshot a recipe row, unconditionally. The repo mints a fresh id per insert
   * and enforces neither limit below — the customer route calls
   * {@link createIfUnderLimit}. Kept for fixtures and internal callers.
   */
  create(args: CreateRecipeArgs): Promise<RecipeRecord>;

  /**
   * Security sweep #7 — the customer save path. Under one per-account lock:
   *
   *   1. A source session is saved ONCE. If the account already has a recipe of
   *      `agentSessionId` with the same label and description, that recipe is the
   *      answer (`existing` — a retried save stores no second copy); with any other
   *      label or description, `session_already_saved`. A save with no source
   *      session (`agentSessionId: null`) is never deduplicated.
   *   2. At most `limit` recipes per account (`limit_reached`).
   *   3. Otherwise the recipe is stored (`created`).
   *
   * v1.0 allowed one session under several labels; each was a full transcript
   * copy, which is what let one session be saved 60 times into 19 MB.
   */
  createIfUnderLimit(args: CreateRecipeArgs & { limit: number }): Promise<CreateRecipeOutcome>;

  /**
   * V-530.I (D2) — list the account's recipes, newest first. Keyset
   * pagination on (createdAt DESC, id DESC), mirroring the prod-proven
   * profiles-repo so same-timestamp rows can't drop at a page boundary.
   * Read-path only; recipe EXECUTION stays gated on the harness executor.
   */
  list(args: ListRecipesArgs): Promise<ListRecipesPage>;

  /**
   * V-530.J (D2) — fetch one recipe, scoped to the account. Returns null
   * when missing OR owned by another account (existence is never leaked
   * cross-account — the route maps null → 404).
   */
  getById(args: { accountId: string; id: string }): Promise<RecipeRecord | null>;

  /**
   * V-530.J (D3) — delete one recipe, scoped to the account. Returns true
   * iff a row was deleted; false = missing or not owned (route → 404).
   */
  deleteById(args: { accountId: string; id: string }): Promise<boolean>;
}

const DEFAULT_RECIPE_PAGE = 50;
const MAX_RECIPE_PAGE = 100;

/**
 * In-memory implementation for unit tests + the disabled-routes
 * activation-gate stub (kept symmetric with AgentSessionsRepo).
 */
export class InMemoryRecipesRepo implements RecipesRepo {
  private readonly rows = new Map<string, RecipeRecord>();

  constructor(private readonly nowFn: () => Date = () => new Date()) {}

  // eslint-disable-next-line @typescript-eslint/require-await
  async create(args: CreateRecipeArgs): Promise<RecipeRecord> {
    const now = this.nowFn();
    const id = `rec_inmem_${randomUUID()}`;
    const validated = validateLabelAndDescription(args.label, args.description);
    const record: RecipeRecord = {
      id,
      accountId: args.accountId,
      agentSessionId: args.agentSessionId,
      label: validated.label,
      description: validated.description,
      intentLog: [...args.intentLog],
      transcriptSnapshot: [...args.transcriptSnapshot],
      createdAt: now,
      updatedAt: now,
    };
    this.rows.set(id, record);
    return record;
  }

  /**
   * Security sweep #7 — mirrors DrizzleRecipesRepo.createIfUnderLimit (the same
   * three rules, in the same order). Sequential, so it needs no lock; the Drizzle
   * method's lock is exercised against Postgres.
   */
  async createIfUnderLimit(
    args: CreateRecipeArgs & { limit: number },
  ): Promise<CreateRecipeOutcome> {
    const validated = validateLabelAndDescription(args.label, args.description);
    if (args.agentSessionId !== null) {
      const saved = [...this.rows.values()]
        .filter((r) => r.accountId === args.accountId && r.agentSessionId === args.agentSessionId)
        .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime() || (a.id < b.id ? -1 : 1));
      const same = saved.find(
        (r) => r.label === validated.label && r.description === validated.description,
      );
      if (same !== undefined) return { kind: 'existing', record: same };
      const first = saved[0];
      if (first !== undefined) return { kind: 'session_already_saved', recipeId: first.id };
    }
    const current = [...this.rows.values()].filter((r) => r.accountId === args.accountId).length;
    if (current >= args.limit) return { kind: 'limit_reached', current };
    return { kind: 'created', record: await this.create(args) };
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  async list(args: ListRecipesArgs): Promise<ListRecipesPage> {
    const limit = Math.min(args.limit ?? DEFAULT_RECIPE_PAGE, MAX_RECIPE_PAGE);
    // (createdAt DESC, id DESC) — same total order as the Drizzle keyset.
    let rows = [...this.rows.values()]
      .filter((r) => r.accountId === args.accountId)
      .sort((a, b) => {
        const t = b.createdAt.getTime() - a.createdAt.getTime();
        if (t !== 0) return t;
        return a.id < b.id ? 1 : a.id > b.id ? -1 : 0;
      });

    if (args.cursor !== undefined) {
      const cur = this.rows.get(args.cursor);
      if (cur !== undefined && cur.accountId === args.accountId) {
        const curT = cur.createdAt.getTime();
        rows = rows.filter(
          (r) => r.createdAt.getTime() < curT || (r.createdAt.getTime() === curT && r.id < cur.id),
        );
      }
    }

    const hasMore = rows.length > limit;
    const data = rows.slice(0, limit);
    const nextCursor =
      hasMore && data.length > 0 ? (data[data.length - 1] as RecipeRecord).id : null;
    return { data, hasMore, nextCursor };
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  async getById(args: { accountId: string; id: string }): Promise<RecipeRecord | null> {
    const row = this.rows.get(args.id);
    return row !== undefined && row.accountId === args.accountId ? row : null;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  async deleteById(args: { accountId: string; id: string }): Promise<boolean> {
    const row = this.rows.get(args.id);
    if (row === undefined || row.accountId !== args.accountId) return false;
    this.rows.delete(args.id);
    return true;
  }
}

function validateLabelAndDescription(
  label: string,
  description: string | undefined,
): { label: string; description: string | null } {
  const trimmedLabel = label.trim();
  if (trimmedLabel.length < 1 || trimmedLabel.length > 120) {
    throw new Error('Recipe label must be 1-120 characters after trim');
  }
  if (description !== undefined && description.length > 2000) {
    throw new Error('Recipe description must be <= 2000 characters');
  }
  return {
    label: trimmedLabel,
    description: description === undefined || description === '' ? null : description,
  };
}

export interface RecipeSuggestion {
  suggestedLabel: string;
  suggestedDescription: string;
}

/**
 * Doc-132 §5.2 (recipe auto-generation) — v1.0 slice. A real
 * cross-customer ML pipeline ("observes patterns, trains a model,
 * feeds published recipes back into training") is out of scope here:
 * it's a genuine customer-data-handling design (Tier 3 per the
 * project's decision-authority policy) that needs a founder call, not
 * a unilateral server-side build. This slice auto-derives a sensible label +
 * description from the CUSTOMER'S OWN intent_log (same data the
 * manual "Save recipe" flow already snapshots for them) so the save
 * dialog prefills something useful instead of a blank form — safe,
 * single-account, no data leaves the account, no training involved.
 *
 * Deterministic (not a model call): scans the ordered intents for the
 * first distinct navigate hostnames + counts interact actions, and
 * composes a short label + one-line description. Never throws —
 * empty/unrecognized logs fall back to a generic label so the caller
 * always gets a usable suggestion.
 */
export function suggestRecipeMetadata(intentLog: ReadonlyArray<AgentIntent>): RecipeSuggestion {
  const hosts: string[] = [];
  for (const intent of intentLog) {
    if (intent.kind !== 'navigate') continue;
    let host: string;
    try {
      host = new URL(intent.url).hostname.replace(/^www\./, '');
    } catch {
      continue; // malformed URL — skip rather than throw
    }
    if (hosts[hosts.length - 1] !== host) hosts.push(host);
  }
  const primaryHost = hosts[0];

  const typeCount = intentLog.filter((i) => i.kind === 'interact' && i.action === 'type').length;
  const tapCount = intentLog.filter((i) => i.kind === 'interact' && i.action === 'tap').length;
  const hasSubmit = intentLog.some((i) => i.kind === 'interact' && i.action === 'press');

  const label = primaryHost
    ? typeCount > 0
      ? `Fill form on ${primaryHost}`
      : `Automation on ${primaryHost}`
    : 'Untitled automation';

  const descriptionParts: string[] = [];
  if (primaryHost) descriptionParts.push(`Navigates to ${primaryHost}`);
  if (typeCount > 0) descriptionParts.push(`fills ${typeCount} field${typeCount === 1 ? '' : 's'}`);
  if (tapCount > 0) descriptionParts.push(`taps ${tapCount} element${tapCount === 1 ? '' : 's'}`);
  if (hasSubmit) descriptionParts.push('submits');
  const description =
    descriptionParts.length > 0
      ? `${descriptionParts.join(', ')}.`
      : `Replays ${intentLog.length} recorded step${intentLog.length === 1 ? '' : 's'}.`;

  return {
    suggestedLabel: label.slice(0, 120),
    suggestedDescription: description.slice(0, 2000),
  };
}
